engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsValueOfTypeNode.java

```diff
@@ -9,155 +17,207 @@
+    boolean doType(
+        Type expectedType, Object payload, @CachedLibrary(limit = "3") TypesLibrary types) {
+      Object tpeOfPayload = typeOfNode.execute(payload);
+      if (profile.profile(isSameObject.execute(expectedType, tpeOfPayload))) {
         return true;
+      } else if (TypesGen.isType(tpeOfPayload)) {
+        Type tpe = TypesGen.asType(tpeOfPayload);
+        Type superTpe = tpe.getSupertype();
+
+        while (tpe != superTpe && superTpe != null) {
+          boolean testSuperTpe = isSameObject.execute(expectedType, superTpe);
+          if (testSuperTpe) {
+            return true;
+          }
+          tpe = superTpe;
+          superTpe = superTpe.getSupertype();
+        }
       }
-      actual = actual.getSupertype();
+      return false;
+    }
+
+    @Specialization(
+        guards = {
+          "isArrayType(expectedType)",
+          "interop.hasArrayElements(payload)",
+        })
+    public boolean doArrayViaType(
+        Object expectedType,
+        Object payload,
+        @CachedLibrary(limit = "3") InteropLibrary interop,
+        @CachedLibrary(limit = "3") TypesLibrary types) {
+      return EnsoContext.get(this).getBuiltins().array() == types.getType(payload);
+    }
+
+    @Fallback
+    boolean doOther(Object expectedType, Object payload) {
+      return false;
+    }
+
+    boolean isAnyType(Object expectedType) {
+      return EnsoContext.get(this).getBuiltins().any() == expectedType;
+    }
+
+    boolean isArrayType(Object expectedType) {
+      return EnsoContext.get(this).getBuiltins().array() == expectedType;
     }
   }
 
-  @Specialization(guards = {"!isArrayType(expectedType)", "!isAnyType(expectedType)"})
-  boolean doType(Type expectedType, Object payload) {
-    Object tpeOfPayload = typeOfNode.execute(payload);
-    if (profile.profile(isSameObject.execute(expectedType, tpeOfPayload))) {
+  abstract static class Untyped extends Node {
+    private @Child IsSameObjectNode isSameObject = IsSameObjectNode.build();
+    private @Child TypeOfNode typeOfNode = TypeOfNode.build();
+
+    abstract boolean execute(Object expectedType, Object payload);
+
+    @Specialization(
+        guards = {
+          "isArrayType(expectedType)",
+          "interop.hasArrayElements(payload)",
+        })
+    boolean doArray(
+        Object expectedType, Object payload, @CachedLibrary(limit = "3") InteropLibrary interop) {
       return true;
-    } else if (TypesGen.isType(tpeOfPayload)) {
-      Type tpe = TypesGen.asType(tpeOfPayload);
-      Type superTpe = tpe.getSupertype();
+    }
 
-      while (tpe != superTpe && superTpe != null) {
-        boolean testSuperTpe = isSameObject.execute(expectedType, superTpe);
-        if (testSuperTpe) {
-          return true;
-        }
-        tpe = superTpe;
-        superTpe = superTpe.getSupertype();
+    @Specialization(guards = {"isMapType(expectedType, iop)", "iop.hasHashEntries(payload)"})
+    boolean doMap(
```

Comment:
I am no longer sure this is good idea. We'd rather concentrate all the logic of checking the type into a single node. I guess I revert changes in this file and just treat everything that `hasHashElements` as `Map` in `TypeOfNode`. While trying to do so I decided to do a bit of cleanup: 
- #6819